server/jobWorkerServer.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/P-A-R-U-S/Go-Job-Worker-Service/pkg/jobWorker"
+	"github.com/P-A-R-U-S/Go-Job-Worker-Service/pkg/proto"
+	"github.com/P-A-R-U-S/Go-Job-Worker-Service/pkg/tls"
+	"google.golang.org/grpc"
+	"io"
+	"sync"
+)
+
+var (
+	ErrJobNotFound   = errors.New("job not found")
+	ErrNotAuthorized = errors.New("user is not authorized to access job")
+)
+
+type userJob struct {
+	user string
+	job  *jobWorker.Job
+}
+
+type JobWorkerServer struct {
+	userJobs map[string]userJob
+	mutex    sync.Mutex
+	rwmutex  sync.RWMutex
+}
+
+func NewJobWorkerServer() *JobWorkerServer {
+	return &JobWorkerServer{
+		userJobs: map[string]userJob{},
+	}
+}
+
+// Start creates a new job for the user and starts the job.
+func (s *JobWorkerServer) Start(ctx context.Context, request *proto.JobCreateRequest) (*proto.JobResponse, error) {
+	user, err := tls.GetUserFromContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get user from certificate: %w", err)
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	config := jobWorker.JobConfig{
+		CPU:              request.CPU,
+		IOBytesPerSecond: request.IoBytesPerSecond,
+		MemBytes:         request.MemBytes,
+		Command:          request.GetCommand(),
+		Arguments:        request.GetArgs(),
+	}
+
+	newJob := jobWorker.NewJob(&config)
+
+	s.userJobs[newJob.UUID.String()] = userJob{
+		user: user,
+		job:  newJob,
+	}
+
+	if err = newJob.Start(); err != nil {
+		return &proto.JobResponse{Id: newJob.UUID.String()}, fmt.Errorf("error starting job: %w", err)
+	}
+
+	return &proto.JobResponse{Id: newJob.UUID.String()}, nil
+}
+
+func (s *JobWorkerServer) Status(ctx context.Context, request *proto.JobRequest) (*proto.JobStatusResponse, error) {
+	jobID := request.GetId()
+
+	job, ok := s.userJobs[jobID]
+	if !ok {
+		return nil, ErrJobNotFound
+	}
+
+	user, err := tls.GetUserFromContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get user from certificate: %w", err)
+	}
+
+	if user != job.user {
+		// TODO: In production to prevent analyze security vulnerabilities
+		// 		 better to returning Not Found instead of Permission Denied to hide job existence
+		return nil, ErrNotAuthorized
+	}
+
+	// we don't want to lock until find a job
+	s.rwmutex.RLock()
+	defer s.rwmutex.RUnlock()
+
+	jobStatus := job.job.Status()
```

Comment:
it's unclear what `s.rwmutex` is protecting here. since `s.rwmutex` belongs to server, it should protect its immediate members like `s.userJob[jobID]`. Currently there is a race when `s.userJobs` is read above. Job's status should be protected by the job itslef, not by the server.

Reply:
@greedy52 thank for the comments. Yes, Job status are protected by job itself:  https://github.com/P-A-R-U-S/Go-Job-Worker-Service/blob/c1b75b7d8a4e36aefe0bfae3196c251d5ff15dca/pkg/jobWorker/job.go#L290

<img width="605" alt="Screenshot 2024-09-10 at 9 33 00 AM" src="https://github.com/user-attachments/assets/a8447c84-9daf-485b-aeef-e7fdb04d2c04">

But using RWMutex let us to unblock Stop (see @smallinsky  comment: https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/3#discussion_r1749755721)